pkg/provisioner/conditionwatcher.go

```diff
@@ -0,0 +1,170 @@
+
+	c := conditionWatcher{
+		informerFactory: informerFactory,
+		waitMap:         sync.Map{},
+	}
+
+	azVolumeAttachmentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+	azVolumeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+	azDriverNodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+
+	go informerFactory.Start(ctx.Done())
+
+	synced := cache.WaitForCacheSync(ctx.Done(), azVolumeAttachmentInformer.HasSynced, azVolumeInformer.HasSynced, azDriverNodeInformer.HasSynced)
+	if !synced {
+		klog.Fatalf("Unable to sync caches for azuredisk CRIs")
+		os.Exit(1)
+	}
+
+	return &c
+}
+
+func (c *conditionWatcher) waitForObjectCondition(ctx context.Context, objectType, objectName string, conditionFunc func(obj interface{}) (bool, error)) (*conditionWaiter, error) {
+	klog.V(5).Infof("Adding a condition function for %s (%s)", objectType, objectName)
+	entry := waitEntry{
+		conditionFunc: conditionFunc,
+		waitChan:      make(chan waitResult, 1),
+	}
+	_, exists := c.waitMap.LoadOrStore(objectName, &entry)
+	if exists {
+		err := status.Errorf(codes.FailedPrecondition, "another wait operation in process for %s (%s)", objectType, objectName)
+		klog.Error(err)
+		return nil, err
+	}
+
+	return &conditionWaiter{
+		objectType: objectType,
+		objectName: objectName,
+		watcher:    c,
+	}, nil
+}
+
+func (c *conditionWatcher) getWaitEntry(objectName string) *waitEntry {
+	v, ok := c.waitMap.Load(objectName)
+	if !ok {
+		klog.Errorf("failed to get wait channel for object (%s)", objectName)
+		return nil
+	}
+	entry := v.(*waitEntry)
+	return entry
+}
+
+func (c *conditionWatcher) onUpdate(_, newObj interface{}) {
+	metaObj := newObj.(metav1.Object)
+	klog.V(5).Infof("update event for obj (%s)", metaObj.GetName())
+	c.handleEvent(newObj)
+}
+
+func (c *conditionWatcher) onDelete(obj interface{}) {
+	metaObj := obj.(metav1.Object)
+	klog.V(5).Infof("delete event for obj (%s)", metaObj.GetName())
+	c.handleEvent(obj)
+}
+
+func (c *conditionWatcher) handleEvent(obj interface{}) {
+	metaObj := obj.(metav1.Object)
+	klog.V(5).Infof("reconciling for an event for object (%s)", metaObj.GetName())
+	v, ok := c.waitMap.Load(metaObj.GetName())
+	if !ok {
+		return
+	}
+	klog.V(5).Infof("found a wait entry for object (%s)", metaObj.GetName())
+	entry := v.(*waitEntry)
+	conditionFunc := entry.conditionFunc
+	waitChan := entry.waitChan
+
+	result := waitResult{}
+
+	ok, err := conditionFunc(obj)
+	klog.V(5).Infof("condition result: succeeded: %v, error: %v", ok, err)
+	// if err found, send error through channel
+	if !ok {
+		result.err = err
+		// if no error was found but condition not met, return
+		if err == nil {
+			return
+		}
+	}
+
+	runtimeObj, ok := obj.(runtime.Object)
+	if ok {
```

Comment:
Should `!ok` produce an error in the `result`? We don't expect this to be the case since all K8s objects should implement this interface, but the caller would likely not expect `result.obj == nil` when `result.err == nil`